sessionmanager.md

```diff
@@ -0,0 +1,161 @@
+  }
+}
+```
+
+Multiple edits or transformations could be sent for a document, at any time.
+
+To handle this the user would need to write code similar to this:
+
+```javascript
+while (true) {
+  // some sort of check to make sure we're not processing
+  // too many sessions at a time. The odds of having more documents
+  // than available processors is _high_.
+  acquireSemaphore();
+
+  // by not specifying a session id we will get the next unlocked session from service bus.
+  const sessionReceiver: SessionReceiver = serviceBusClient.getSessionReceiver(
+    "queue",
+    "peekLock"
+  );
+  startSessionLockAutoRenewal(sessionReceiver);
+
+  sessionReceiver.subscribe(messageHandlers);
+
+  // ...after some predefined period of time
+  sessionReceiver.close();
+
+  stopSessionLockRenewal(sessionReceiver);
+  releaseSemaphore();
+}
+```
+
+## Proposal
+
+In the same vein as what we provide in `Receiver.subscribe()` we can provide a
+`SessionManager` that would support the use case where the user would like
+to read messages, in parallel, from a number of sessions while controlling
+the amount of parallelism.
+
+This would simplify session usage to the point where it becomes more accessible,
+allowing more customers to use this powerful feature.
+
+The proposed interface for this change:
+
+```javascript
+class ServiceBusClient {
+  // a session manager is conceptually a receiver, but does not have to
+  // obey the same interface.
+  createSessionManager("queue", "peekLock", sessionManagerOptions): SessionManager;
+  createSessionManager("topic", "subscription", "peekLock", sessionManagerOptions): SessionManager;
+}
+
+interface SessionManagerOptions {
+  // amount of time we will wait for messages to arrive in a session before
+  // it is `close()`d
+  idleTimeoutMs: number;
+
+  // the maximum number of receivers that can be active at any
+  // time. Allows the customer to avoid getting steamrolled by too many
+  // sessions.
+  maxConcurrentReceivers: number;
+
+  // new feature (and probably pushed down to SessionReceiver as well) to enable auto-renewal
+  // of the session lock.
+  maxSessionAutoRenewLockDurationInMs?: number;
+
+  // quite a few of the options available in MessageHandler (used for normal Receiver subscribe)
+  // would also apply here.
+  autoComplete?: boolean;
+}
+
+// implicit to this - this is specific to a single session.
+interface SessionContext {
+  // managing session state is implicit to the context where it is called. This flows
+  // similarly to what developers would do with `SessionReceiver`
+  setState();
+  getState();
+
+  sessionId: string;
+}
+
+interface SessionMessageHandler {
+  // tentatively suggesting proposal #2 for error handling below.
+  processMessage(message: ReceivedMessage|ReceivedMessageWithLock, context: SessionContext);
+  processError(err: Error, context: { sessionId: string });
+}
+
+interface SessionManager {
+  // the handlers
+  subscribe(handlers: SessionMessageHandler);
```

Comment:
SessionManager manager = ServiceBusClient.createSessionManager ( myQueue, sessionManagerOptions );
manager .subscribe(myHandler);

In above code how does user  start and stop/pause receiving the message ?

Reply:
There's a separate item about "stopping/pausing" that we we're discussing right now (it's part of this PR as setting_messages_after_receiver_is_closed.md).

It's a larger issue that affects all of our subscription models (and even track 1).